azurerm/internal/services/postgres/postgresql_server_key_resource.go

```diff
@@ -0,0 +1,193 @@
+package postgres
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/mgmt/2018-02-14/keyvault"
+	"github.com/Azure/azure-sdk-for-go/services/postgresql/mgmt/2020-01-01/postgresql"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/locks"
+	keyVaultParse "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/keyvault/parse"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/postgres/parse"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/postgres/validate"
+	azSchema "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmPostgreSQLServerKey() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmPostgreSQLServerKeyCreateUpdate,
+		Read:   resourceArmPostgreSQLServerKeyRead,
+		Update: resourceArmPostgreSQLServerKeyCreateUpdate,
+		Delete: resourceArmPostgreSQLServerKeyDelete,
+
+		Importer: azSchema.ValidateResourceIDPriorToImport(func(id string) error {
+			_, err := parse.PostgreSQLServerKeyID(id)
+			return err
+		}),
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"server_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.PostgresServerServerID,
+			},
+
+			"key_vault_key_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: azure.ValidateKeyVaultChildId,
```

Comment:
Should this also support the "versionless" secrets? If so we'll likely need to look up the current version to be able to use that as a name?

Reply:
By saying `key_vault_key_id` here, it means some key URL WITH version like this:
```
https://YourVaultName.vault.azure.net/keys/YourKeyName/01234567890123456789012345678901>
```
Do you suggest we change this schema to align with the CMK for storage? i.e, like this:
```
key_vault_id = 
key_name = 
key_version = 
```
to emphasize that we are requiring a versioned key here?

Reply:
right however most Azure API's which support a versioned secret also take a "versionless" version too (since the 'version' gets parsed as empty, which the Key Vault API's return as "latest") - so we should test that

No, this makes sense to leave as `key_vault_key_id` - however we need to confirm if this validation needs to accept both a versioned and versionless Key ID, which we'll identify through testing that?

Reply:
Sure I will have a try with this

Reply:
Just have a try if we pass a key URL without version, and get an `internalServerError`.
And per their [document page](https://docs.microsoft.com/en-us/azure/postgresql/howto-data-encryption-cli#set-data-encryption-for-azure-database-for-postgresql-single-server), they should be expecting the key URL with version

Reply:
Then should we be validing that the id has a version then?

Reply:
This function `ValidateKeyVaultChildId` is expecting this ID to have a version. We have a `ValidateKeyVaultChildIdVersionOptional` for the version is optional scenario